databricks/koalas/typedef/typehints.py

```diff
@@ -93,41 +93,48 @@ class NameTypeHolder(object):
 
 def as_spark_type(tpe) -> types.DataType:
     """
-    Given a python type, returns the equivalent spark type.
+    Given a Python type, returns the equivalent spark type.
     Accepts:
-    - the built-in types in python
+    - the built-in types in Python
     - the built-in types in numpy
     - list of pairs of (field_name, type)
     - dictionaries of field_name -> type
-    - python3's typing system
+    - Python3's typing system
     """
-    if tpe in (str, "str", "string"):
-        return types.StringType()
-    elif tpe in (bytes,):
+    # ArrayType
+    if tpe in (np.ndarray,):
+        # TODO: support other child types
+        return types.ArrayType(types.StringType())
+    # BinaryType
+    elif tpe in (bytes, np.character):
         return types.BinaryType()
+    # BooleanType
+    elif tpe in (bool, np.bool, "bool", "boolean"):
+        return types.BooleanType()
+    # DateType
+    elif tpe in (datetime.date,):
+        return types.DateType()
+    # NumericType
     elif tpe in (np.int8, "int8", "byte"):
         return types.ByteType()
-    elif tpe in (np.int16, "int16", "short"):
-        return types.ShortType()
-    elif tpe in (int, "int", np.int, np.int32):
-        return types.IntegerType()
-    elif tpe in (np.int64, "int64", "long", "bigint"):
-        return types.LongType()
-    elif tpe in (float, "float", np.float):
-        return types.FloatType()
-    elif tpe in (np.float64, "float64", "double"):
-        return types.DoubleType()
     elif tpe in (decimal.Decimal,):
         return types.DecimalType(38, 18)
-    elif tpe in (datetime.datetime, np.datetime64):
+    elif tpe in (float, np.float, np.float64, "float", "float64", "double"):
+        return types.DoubleType()
+    elif tpe in (np.float32, "float32"):
+        return types.FloatType()
+    elif tpe in (np.int32, "int32"):
+        return types.IntegerType()
+    elif tpe in (int, np.int, np.int64, "int", "int64", "long", "bigint"):
+        return types.LongType()
+    elif tpe in (np.int16, "int16", "short"):
+        return types.ShortType()
```

Comment:
Why did you change the order of `if-else`s?

Reply:
I re-order them according to the order of the class hierarchy described in the [spark documentation](https://spark.apache.org/docs/3.0.0-preview/api/java/org/apache/spark/sql/types/package-tree.html). If this is not necessary, will revert it.

Reply:
ah, I see.